### PR TITLE
Mlflow logger

### PR DESCRIPTION
**Pull request by y-ksenia**
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .`? 
- [x] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Description

New logger for MLflow.

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/loggers/functional.py [hunk 1]
```diff
@@ -14,3 +14,64 @@
...
+    "runner",
+    "engine",
+    "model",
+    "stages",
+)
+STAGE_PARAMS = ("data", "criterion", "optimizer", "scheduler", "stage")
+EXCLUDE_PARAMS = ("loggers", "transform", "callbacks")
+
+
+def mlflow_log_dict(dictionary: Dict[str, Any], prefix: str = "", log_type: Optional[str] = None):
```
Comment (Scitator) [inline catalyst/loggers/functional.py]: it it's only mlflow-based, could we move it to `catalyst/loggers/mlflow.py`?

File: catalyst/runners/config.py [hunk 2]
```diff
@@ -19,6 +19,7 @@
 from catalyst.engines import IEngine
 from catalyst.loggers.console import ConsoleLogger
 from catalyst.loggers.csv import CSVLogger
+from catalyst.loggers.mlflow import MlflowLogger
```
Comment (Scitator) [inline catalyst/runners/config.py]:
```suggestion
from catalyst.settings import SETTINGS
if SETTINGS.mlflow_required:
    from catalyst.loggers.mlflow import MlflowLogger
```

File: catalyst/runners/runner.py [hunk 3]
```diff
@@ -17,6 +17,7 @@
 from catalyst.engines import DeviceEngine, IEngine
 from catalyst.loggers.console import ConsoleLogger
 from catalyst.loggers.csv import CSVLogger
+from catalyst.loggers.mlflow import MlflowLogger
```
Comment (Scitator) [inline catalyst/runners/runner.py]: why do we need it here?

File: catalyst/core/runner.py [hunk 4]
```diff
@@ -527,6 +527,7 @@ def on_stage_start(self, runner: "IRunner"):
         self._setup_loaders()
         self._setup_components()
         self._setup_callbacks()
+        self.log_hparams(hparams=self.hparams, scope=runner.stage_key)
```
Comment (Scitator) [inline catalyst/core/runner.py]:
```suggestion
        self.log_hparams(hparams=self.hparams, scope="stage")
```

File: catalyst/loggers/__init__.py [hunk 5]
```diff
@@ -3,3 +3,4 @@
 from catalyst.loggers.console import ConsoleLogger
 from catalyst.loggers.csv import CSVLogger
 from catalyst.loggers.tensorboard import TensorboardLogger
+from catalyst.loggers.mlflow import MlflowLogger
```
Comment (Scitator) [inline catalyst/loggers/__init__.py]:
```suggestion
from catalyst.settings import SETTINGS
if SETTINGS.mlflow_required:
    from catalyst.loggers.mlflow import MlflowLogger
```

File: catalyst/runners/config.py (hunk 2 above)
Comment (Scitator) [reply catalyst/runners/config.py]: why do we need it?

File: catalyst/runners/hydra.py [hunk 6]
```diff
@@ -20,6 +20,7 @@
 from catalyst.engines import IEngine
 from catalyst.loggers.console import ConsoleLogger
 from catalyst.loggers.csv import CSVLogger
+from catalyst.loggers.mlflow import MlflowLogger
```
Comment (Scitator) [inline catalyst/runners/hydra.py]: and here

File: catalyst/loggers/mlflow.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+from catalyst.core.logger import ILogger
+from catalyst.loggers.functional import (
+    EXCLUDE_PARAMS,
+    EXPERIMENT_PARAMS,
+    mlflow_log_dict,
+    STAGE_PARAMS,
+)
+
+
+class MlflowLogger(ILogger):
```
Comment (Scitator) [inline catalyst/loggers/mlflow.py]: could you please add it to the docs? https://github.com/catalyst-team/catalyst/blob/dev/docs/api/loggers.rst

Comment (Scitator) [reply catalyst/loggers/mlflow.py]: and could you please write some tests for it? for example, you could add it to our `test_finetuning` pipeline https://github.com/catalyst-team/catalyst/blob/dev/catalyst/tests/test_finetune2.py#L29

File: catalyst/loggers/functional.py (hunk 1 above)
Comment (Scitator) [reply catalyst/loggers/functional.py]: up
let's keep all mlfow-related functions into `catalyst/loggers/mlflow` :)

File: catalyst/loggers/mlflow.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+                _mlflow_log_dict(exp_params, log_type="param")
+
+    def close_log(self) -> None:
+        """Finds all **running** runs and ends them."""
+        all_runs = mlflow.search_runs()
+        for _ in all_runs[all_runs.status == "RUNNING"]:
+            mlflow.end_run()
+
+
+__all__ = ["MLflowLogger", "EXPERIMENT_PARAMS", "STAGE_PARAMS", "EXCLUDE_PARAMS"]
```
Comment (Scitator) [inline catalyst/loggers/mlflow.py]:
```suggestion
__all__ = ["MLflowLogger"]
```

File: catalyst/loggers/mlflow.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        registry_uri: Optional[str] = None,
+    ) -> None:
+        """
+        Args:
+            experiment: Name of the experiment in MLflow to log to.
+            run: Name of the run in MLflow to log to.
+            tracking_uri: URI of tracking server against which
+                to log run information related.
+            registry_uri: Address of local or remote model registry server.
+        """
```
Comment (Scitator) [inline catalyst/loggers/mlflow.py]: could you please move it to the class description?
example: https://github.com/catalyst-team/catalyst/blob/dev/catalyst/metrics/_accuracy.py#L12#L34

Comment (y-ksenia) [reply catalyst/loggers/mlflow.py]: done

File: catalyst/loggers/mlflow.py (hunk 8 above)
Comment (y-ksenia) [reply catalyst/loggers/mlflow.py]: 👌

File: catalyst/loggers/mlflow.py (hunk 7 above)
Comment (y-ksenia) [reply catalyst/loggers/mlflow.py]: done

File: catalyst/contrib/utils/visualization.py [hunk 10]
```diff
@@ -2,28 +2,11 @@
 
 import matplotlib
 import numpy as np
-import torch
 
 matplotlib.use("Agg")
 import matplotlib.pyplot as plt  # noqa: E402
```
Comment (Scitator) [inline catalyst/contrib/utils/visualization.py:8]:
```suggestion
```

File: catalyst/loggers/functional.py [hunk 11]
```diff
@@ -1,3 +1,7 @@
+import numpy as np
+import torch
+
+
```
Comment (Scitator) [inline catalyst/loggers/functional.py:4]:
```suggestion
```

**Pull request merged**